### PR TITLE
Move tensor->flow passes to Dialect/Flow/Conversion/TensorToFlow.

### PR DESCRIPTION
**Pull request by ScottTodd**
General cleanup and the `tensor`->`flow` conversions will be useful outside of `iree/compiler/InputConversion/Common/`, particularly for detensoring (https://github.com/google/iree/issues/1159).

* Moved all patterns in `iree/compiler/InputConversion/Common/ConvertUpstreamToIREE.cpp` into `iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp` and folded that pass into `IREE::Flow::createConvertToFlowTensorOpsPass`
* Moved `tensor`->`flow` patterns and their tests from `iree/compiler/Dialect/Flow/Transforms/` into the new `Flow/Conversion/TensorToFlow` directory. Changed them from `OpRewritePattern`s to `OpConversionPattern`s. `linalg`->`flow` passes remain in their original location
* Updated pattern names, test case names, comments, etc. following the renaming in https://github.com/llvm/llvm-project/commit/060208b4c8b78b2456b8440d9597c9f584676bf4 / https://github.com/google/iree/commit/7f89eb64ee471895fe45493061c61b9cdca79194

### CODE REVIEW

File: iree/tools/init_iree_passes.h [hunk 1]
```diff
@@ -46,6 +47,7 @@ inline void registerAllIreePasses() {
 
   linalg_ext::registerLinalgExtPasses();
   IREE::Flow::registerFlowPasses();
+  IREE::Flow::registerTensorToFlowPassPipelines();
```
Comment (ScottTodd) [inline iree/tools/init_iree_passes.h:50]: @benvanik I was wondering if the new pass should be registered in `registerFlowPasses` instead of adding a new function. WDYT?

Comment (benvanik) [reply iree/tools/init_iree_passes.h:50]: ah, we don't want a pass for this here at all - we just want patterns that can be added to other passes
see for example:
https://github.com/google/iree/blob/552d3f888e46f96c8227b1bd2e4600b85338237b/iree/compiler/Dialect/HAL/Conversion/FlowToHAL/ConvertFlowToHAL.h
https://github.com/google/iree/blob/552d3f888e46f96c8227b1bd2e4600b85338237b/iree/compiler/Dialect/VM/Conversion/MemRefToVM/ConvertMemRefToVM.h

We want to add these patterns to the existing https://github.com/google/iree/blob/323108e3869648a42dbfc82f8eb54cf05bdbafef/iree/compiler/Dialect/Flow/Transforms/ConvertToFlowTensorOps.cpp (you'll note it already ended up growing some tensor->flow patterns that should be moved to your new Conversions/TensorToFlow/ directory)

so: delete all the pass stuff, replace with a populate patterns method, call that method from ConvertToFlowTensorOps

**Review by benvanik — CHANGES_REQUESTED** [review]

File: iree/tools/init_iree_passes.h (hunk 1 above)
Comment (ScottTodd) [reply iree/tools/init_iree_passes.h:50]: Gotcha. That's the kind of feedback I was hoping for, since there's a mix of different styles depending on when compiler code was written or by whom >_>

I hesitated to only provide patterns, since there's quite a bit of boilerplate in the existing pass that I'm moving: https://github.com/google/iree/blob/1cd9f0ab6995d0333f7d76412a4bfdcc3c5d35b6/iree/compiler/InputConversion/Common/ConvertUpstreamToIREE.cpp#L135-L157

Comment (benvanik) [reply iree/tools/init_iree_passes.h]: The FlowToHAL has the setupFlowToHALLegality method that does those ConversionTarget manipulations - you can do the same here so that the tensor CastOp/FromElements/etc can be made illegal as needed.

Comment (ScottTodd) [reply iree/tools/init_iree_passes.h]: Still poking at this, but I think I'm moving in the right direction.

I feel like there should be a way to not have the conversion patterns and "regular" patterns separate. (might be not understanding the terminology + MLIR infra, or maybe some of the patterns should be converted to one or the other)

Comment (ScottTodd) [reply iree/tools/init_iree_passes.h]: Made those changes, but still working out a few test failures (since this now applies more patterns in several places).

File: iree/compiler/Dialect/Flow/Transforms/ConvertToFlowTensorOps.cpp [hunk 2]
```diff
@@ -292,12 +110,27 @@ struct ConvertToFlowTensorOpsPass
     FuncOp funcOp = getOperation();
     MLIRContext *context = funcOp->getContext();
     context->allowUnregisteredDialects(true);
+
+    // Run partial conversion from tensor -> flow.tensor ops.
```
Comment (ScottTodd) [inline iree/compiler/Dialect/Flow/Transforms/ConvertToFlowTensorOps.cpp:114]: Seems like this refactoring is changing some preconditions that the `linalg_ext.sort` lowering relied on (for dynamic shapes?), but I'm not sure yet what the specific issue is. This change should only be moving patterns around, it does not change the details within any patterns.

Before:
* ConvertUpstreamToIREE for `tensor.cast` and `tensor.from_elements`
* ConvertToFlowTensorOps for `tensor.insert_slice` and `tensor.extract_slice`

After:
* ConvertToFlowTensorOps for `tensor.cast`, `tensor.from_elements`, `tensor.insert_slice`, and `tensor.extract_slice`

Previously `memref::createResolveShapedTypeResultDimsPass` was removing `tensor.cast` ops, but now they're already converted to `flow.tensor.reshape`

@hanhanW or @MaheshRavishankar any ideas? I can make this refactor more of an NFC, but organizing all tensor->flow patterns together seemed like a good idea to me. Detensoring will also want the `tensor.from_elements` patterns, so I at least want to put those in a spot other than `ConvertUpstreamToIREE`

Comment (ScottTodd) [reply iree/compiler/Dialect/Flow/Transforms/ConvertToFlowTensorOps.cpp:114]: I had a chat with Mahesh offline.

* The pass should not be converting `tensor.cast` ops into `flow.tensor.reshape` ops within dispatch regions - flow ops should only exist outside of dispatch regions.
* I could split the patterns to get this change back to an NFC (make `createConvertToFlowTensorOpsPass(true)` and `createConvertToFlowTensorOpsPass(false)` match their prior behavior, expose patterns that were in `ConvertUpstreamToIREE` through a different mechanism)
* Writing lit tests for the `// Elementwise, fusion, tiling and distribution.` block of passes would be tricky. Lots of optional optimizations and specific sequencing. Maybe we could verify no flow ops within dispatch regions though? (I'd like some more specific error message for my incorrect pass compared to this e2e sort.mlir test failing and needing to dig through why)

Comment (benvanik) [reply iree/compiler/Dialect/Flow/Transforms/ConvertToFlowTensorOps.cpp:114]: (no [magic booleans](https://blog.codinghorror.com/avoiding-booleans/) - have a different pass name or an enum or something meaningful)

Comment (ScottTodd) [reply iree/compiler/Dialect/Flow/Transforms/ConvertToFlowTensorOps.cpp:114]: Oh, the booleans came from https://github.com/google/iree/pull/5614 :P 

I'd be doing something more like this for the second bullet point
```diff
void populateTensorToFlowPatterns(MLIRContext *context,
                                  OwningRewritePatternList &patterns,
                                  TypeConverter &typeConverter);

+void populateDispatchTensorToFlowPatterns(MLIRContext *context,
+                                          OwningRewritePatternList &patterns,
+                                          TypeConverter &typeConverter);
```

Going to first see if only operating outside of dispatch regions (the first bullet) is sufficient.

Comment (ScottTodd) [reply iree/compiler/Dialect/Flow/Transforms/ConvertToFlowTensorOps.cpp:114]: Nice, I think that change worked, no need to split the patterns:

```c++
  static bool shouldBeConverted(tensor::CastOp op) {
    if (op->getParentOfType<Flow::DispatchWorkgroupsOp>()) {
      return false;
    }

    return true;
  }
```

I'll get this ready for review next week. I should add some unit tests for those `shouldBeConverted` checks. It looks like _some_ flow ops can exist within dispatch regions, but maybe a selective verifier could still help spot issues like this earlier?

**Comment by ScottTodd** [comment]
(The commit history isn't well organized, LMK if you'd prefer for me to split this into a more cohesive set of commits for review and/or non-squash merge)

File: iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+  SmallVector<Value, 4> dynamicDims;
+  for (auto dim : llvm::enumerate(v.getType().cast<ShapedType>().getShape())) {
+    if (dim.value() != ShapedType::kDynamicSize) continue;
+    dynamicDims.push_back(b.createOrFold<tensor::DimOp>(loc, v, dim.index()));
+  }
+  return dynamicDims;
+}
+
+/// Convert tensor.insert_slice ops into flow.tensor.update ops where possible.
+struct ConvertTensorInsertSlicePattern
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp:124]: Why are these changed to a conversion pattern rewrite? AFAIK, you can still keep this as a `OpRewritePattern` and use it in the dialect conversion. In general better to avoid dialect conversion unless you absolutely need it (and you only absolutely need it if you are changing types, which is not the case here since all types are `tensor`s)

File: iree/compiler/Dialect/Flow/Transforms/ConvertToFlowTensorOps.cpp [hunk 4]
```diff
@@ -292,12 +110,27 @@ struct ConvertToFlowTensorOpsPass
...
+    OwningRewritePatternList conversionPatterns(&getContext());
+    TypeConverter typeConverter;
+    typeConverter.addConversion([](Type t) { return t; });
+    ConversionTarget target(*context);
+    setupTensorToFlowLegality(context, target, typeConverter);
+    populateTensorToFlowPatterns(&getContext(), conversionPatterns,
+                                 typeConverter);
+    if (failed(applyPartialConversion(getOperation(), target,
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/ConvertToFlowTensorOps.cpp:122]: This pass flow is strange. It is first doing dialect conversion , and then doing pattern rewrites. Made me go back and see the changes in `ConvertTensorToFlow` and see that these are dialect conversions. Maybe making the patterns `OpRewritePatterns` and can avoid doing all the dialect conversion boiler plate, and you can just add these patterns to run "before dispatch region" formation case. Then you will have a single `applyPatternsAndFoldGreedily`

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp (hunk 3 above)
Comment (ScottTodd) [reply iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp:124]: Ah, I'll try changing them all back to `OpRewritePatterns`. I wasn't really sure which parts were load bearing in the existing implementations.

Comment (ScottTodd) [reply iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp:124]: The `CastOp` conversion was using dialect conversion features and goes from static<->dynamic dimensions: 
https://github.com/google/iree/blob/3554868e9945cb48fe1ef2b2adfec96253c40171/iree/compiler/InputConversion/Common/ConvertUpstreamToIREE.cpp#L26-L41

Is conversion actually needed there or is this possible to express using an `OpRewritePattern`?

Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp:124]: ah, thats painful. Typically to switch between static <-> dynamic you would use a `tensor.cast` but that doesnt work for `tensor.cast` itself.

Maybe we dont move the pattern into a common place? So `tensor.cast` -> whatever for input conversion stays as is. What is your need for `tensor.cast` conversion?

Comment (ScottTodd) [reply iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp:123]: Hmm.. well, I pushed a commit that seems to work changing back to `OpRewritePattern` for all of these. I don't have a particular use for the `tensor.cast` conversion at this stage, but would prefer to group it with the other `tensor`->`flow` dialect patterns if possible.

Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp:123]: Ok, ill take a look. If keeping it as `OpRewritePattern` doesnt hurt, then its ok.

**Review by MaheshRavishankar — APPROVED** [review]
If this passes all tests, then this is fine.

**Review by benvanik — APPROVED** [review]
